core/src/com/unciv/models/ruleset/nation/Nation.kt

```diff
@@ -257,18 +258,23 @@ class Nation : RulesetObject() {
             }
         }
     }
-    
-    fun matchesFilter(filter: String): Boolean {
-        return MultiFilter.multiFilter(filter, ::matchesSingleFilter)
-    }
+
+    @Transient
+    private val cachedMatchesFilterResult = HashMap<String, Boolean>()
+
+    fun matchesFilter(filter: String, state: StateForConditionals? = null): Boolean =
+        MultiFilter.multiFilter(filter, {
+            cachedMatchesFilterResult.getOrPut(it) { matchesSingleFilter(it) } ||
+                (state != null && hasUnique(it, state) || state == null && hasTagUnique(it))
```

Comment:
All of these "or" clauses have 3 sections, there's no reason to separate the last two with parentheses 

```suggestion
            cachedMatchesFilterResult.getOrPut(it) { matchesSingleFilter(it) }
                || state != null && hasUnique(it, state)
                || state == null && hasTagUnique(it)
```

Reply:
I can handle this tomorrow, but unnecessary nitpick fun fact:
This compiles
```
a ||
b
```
This compiles
```
a +
b
```
This does not compile
```
a
+ b
```
Idk why Idea prefers to write it this way and just wrap the whole thing with parenthesis, but personally, I've started preferring to not put my operators on the new line if it can be avoided for consistency